### PR TITLE
Integrate Login screen with API

### PR DESCRIPTION
**Pull request by longlh-non**
Done:
 - [x] Configure the login API endpoint in the application.
 - [x] Set up requests to submit user credentials (email and password).
 - [x] Handle API responses, including success.
 - [x] Add any state management

To do:
 - [x] Handle API error, edge cases (e.g., incorrect credentials, server downtime).
 - [ ]  Code optimisation.
 
## Has This Been Tested?

- [ ] Yes

_Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration_

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: frontend/api/authAPI.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import baseAPI from "./baseAPI";
+
+export const loginAPI = async (credentials) => {
+  return await baseAPI.post("/auth/login", credentials, {
+    headers: {
+      "Content-Type": "multipart/form-data", // Required for file uploads
```
Comment (Salamehar) [inline frontend/api/authAPI.js:6]: * Login credentials typically shouldn't be sent as multipart/form-data unless you're specifically handling file uploads during login.
* For regular login, you should use application/json.
* No error handling is implemented.

File: frontend/api/baseAPI.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+// Add a request interceptor (e.g., attach token)
+baseAPI.interceptors.request.use(
+  (config) => {
+    // Check if the request URL is in the whitelist
+    const isAuthRoute = AUTH_LIST.some((route) =>
+      config.url.includes(route)
+    );
+    if (!isAuthRoute) {
+      const token = document
+        .cookie("access_token")
```
Comment (Salamehar) [inline frontend/api/baseAPI.js:23]: * Should use document.cookie as a string and parse it properly. Consider using js-cookie instead.

File: frontend/api/baseAPI.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import axios from "axios";
+
+// Define public routes (e.g., login, register)
+const AUTH_LIST = ["/auth/login"];
+
+// Create a reusable Axios instance
+const baseAPI = axios.create({
+  baseURL: process.env.REACT_APP_API_URL || "http://localhost:8000",
```
Comment (Salamehar) [inline frontend/api/baseAPI.js:8]: * Having a hardcoded fallback URL might lead to accidental production hitting localhost. Make it required or have environment specific defaults. (Later)

File: frontend/pages/Demo.js [hunk 4]
```diff
@@ -0,0 +1,260 @@
+// src/pages/Demo.js
```
Comment (Salamehar) [inline frontend/pages/Demo.js:1]: * Too many useState hooks in a single component.
* No error handling for missing or invalid data.

File: frontend/pages/Login/Login.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+const Login = () => {
+  const [username, setUsername] = useState("");
+  const [role, setRole] = useState("");
+  const [password, setPassword] = useState("");
+  const dispatch = useDispatch();
+  const { loading, error } = useSelector((state) => state.auth);
+
+  const handleSubmit = (e) => {
+    e.preventDefault();
+    dispatch(login({ username, password }));
```
Comment (Salamehar) [inline frontend/pages/Login/Login.js]: * No input validation before sending credentials.
* No password strength requirements.
* Role selection is available but not used in login request.
* Missing CSRF protection.
* No proper error handling or display

File: frontend/store/certificates/certificateSlice.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import { createSlice } from "@reduxjs/toolkit";
+import { fetchCertificates } from "./certificateThunk";
+
+const certificateSlice = createSlice({
+  name: "certificates",
+  initialState: {
+    products: [],
```
Comment (Salamehar) [inline frontend/store/certificates/certificateSlice.js]: Naming inconsistency - using 'products' for certificates.

**Review by Salamehar — COMMENTED** [review]
Minor suggestions.

Overall:
* Token handling is insecure (direct cookie manipulation)
* Missing CSRF protection
* Missing proper session management
* Missing HTTP-Only flag in cookies
* Hardcoded API URL fallback (localhost:8000)
* Missing proper error handling in API calls
* Missing request/response validation
* Missing type definitions
* Missing separation of concerns

File: frontend/api/baseAPI.js (hunk 2 above)
Comment (Salamehar) [reply frontend/api/baseAPI.js:23]: Later

File: frontend/pages/Demo.js (hunk 4 above)
Comment (Salamehar) [reply frontend/pages/Demo.js:1]: Later

File: frontend/pages/Login/Login.js (hunk 5 above)
Comment (ClemAtt) [reply frontend/pages/Login/Login.js]: If you can address just these for now, rest later:

> Missing CSRF protection.
> No proper error handling or display

File: frontend/pages/Login/Login.js [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                    marginBottom: 12,
+                    borderColor: "#DADCE0",
+                    color: "#5F6368",
+                  }}
+                >
+                  Don’t have an account?
+                </Divider>
+                <Text>
+                  <Link
+                    href="#"
```
Comment (ClemAtt) [inline frontend/pages/Login/Login.js:158]: Please add the google form link here (open in new tab) https://docs.google.com/forms/d/e/1FAIpQLSdSkHMAYSu43VJFevngfVT5hvnWRZvwkelIf9QaPtpLVrIlxA/viewform?usp=sf_link

**Review by ClemAtt — CHANGES_REQUESTED** [review]
A few changes to make. Some comments don't need to be addressed now and have been flagged with `later`

File: frontend/api/authAPI.js (hunk 1 above)
Comment (longlh-non) [reply frontend/api/authAPI.js:6]: Screenshot 2025-01-13 124600

Hi, 

I tried application/json however the backend gave 422 error code because backend requires `OAuth2PasswordRequestForm ` which strictly requires a form data. Please let me know if I can  make a change to backend code to work on this or keep using `multipart/form-data` or wait for someone handle this for me in backend.

And I'm implementing error handling at the moment.

Please let me know.

Cheers

File: frontend/store/certificates/certificateSlice.js (hunk 6 above)
Comment (longlh-non) [reply frontend/store/certificates/certificateSlice.js]: Done

File: frontend/pages/Login/Login.js (hunk 5 above)
Comment (longlh-non) [reply frontend/pages/Login/Login.js]: Hi, I can't find any CSRF prevention implemented in BE code.

**Review by ClemAtt — APPROVED** [review]

File: frontend/pages/Login/Login.js (hunk 5 above)
Comment (longlh-non) [reply frontend/pages/Login/Login.js]: I made changes for error handling and display

> If you can address just these for now, rest later:
> 
> > Missing CSRF protection.
> > No proper error handling or display
